keras_nlp/metrics/perplexity.py

```diff
@@ -0,0 +1,142 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Perplexity metric implementation based on `keras.metrics.Metric`."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+class Perplexity(keras.metrics.Metric):
+    """Perplexity metric.
+    This class implements the perplexity metric. In short, this class calculates
+    the cross entropy loss and takes its exponent.
+    Note: This implementation is not suitable for fixed-size windows.
+    Args:
+        name: string. Name of the metric instance.
+        dtype: string or tf.dtypes.Dtype. Precision of metric computation. If
+               not specified, it defaults to tf.float32.
+        from_logits: bool. If True, `y_pred` (input to `update_state()`) should
+            be the logits as returned by the model. Otherwise, `y_pred` is a
+            tensor of probabilities.
+        pad_token_id: int. Token ID of the padding token. If provided, the mask
+            is computed by this class (all padding tokens are masked while
+            computing the cross entropy loss). Note that if this field is
+            provided, the `sample_weight` field in `update_state()` is ignored.
+        **kwargs: Other keyword arguments.
+    Examples:
+    ```python
+    # 1. update_state() and result()
+    perplexity = keras_nlp.metrics.Perplexity(name="perplexity")
+    target = tf.experimental.numpy.random.randint(low=0, high=10, size=(2, 5))
+    logits = tf.random.uniform(shape=(2, 5, 10))
+    # 1.1 sample_weight not specified.
+    perplexity.update_state(target, logits)
+    print(perplexity.result())
+    # 1.2 sample_weight specified.
+    # mask token 0
+    sample_weight = tf.cast(tf.math.logical_not(tf.equal(y_true, 0)),
+                            tf.float32)
+    perplexity.update_state(target, logits, sample_weight)
+    print(perplexity.result())
+    # 2. Call perplexity directly.
+    perplexity = keras_nlp.metrics.Perplexity(name="perplexity")
+    print(perplexity(target, logits))
+    # 3. Provide the padding token ID and let the class compute the mask on its
+    # own.
+    perplexity = keras_nlp.metrics.Perplexity(name="perplexity", pad_token_id=0)
+    print(perplexity(target, logits))
+    ```
+    """
+
+    def __init__(
+        self,
+        name="perplexity",
+        dtype=None,
+        from_logits=False,
+        pad_token_id=None,
+        **kwargs,
+    ):
+        super(Perplexity, self).__init__(name=name, dtype=dtype, **kwargs)
+        if dtype is None:
+            self._dtype = tf.float32
+        else:
+            if not dtype.is_floating:
+                raise ValueError(
+                    "`dtype` must be a floating point type. "
+                    f"Received: dtype={dtype}"
+                )
+
+        self.cross_entropy_loss = keras.losses.SparseCategoricalCrossentropy(
```

Comment:
This should be a private member, since it is not directly passed from constructor argument.